examples/react-full-v1/pages/star-wars/@id/+onBeforeRender.tsx

```diff
@@ -2,15 +2,15 @@
 
 import fetch from 'cross-fetch'
 import { filterMovieData } from '../filterMovieData'
-import type { Config, PageContextServer } from 'vike/types'
+import type { Config, PageContextClient, PageContextServer } from 'vike/types'
 import type { MovieDetails } from '../types'
 import { render } from 'vike/abort'
 import React from 'react'
 
 const onBeforeRender: Config['onBeforeRender'] = async (
-  pageContext: PageContextServer
+  pageContext: PageContextServer | PageContextClient
```

Comment:
@brillout we touched on this in the past but for some reason, this just works although `Config['onBeforeRender']` is defined to accept only `PageContextServer`. Do you want me to dig more?

Reply:
My (wild) guess/intuition is that it's an inversion of control thing: Vike calls the hook and always provides `pageContext: PageContextServer`, which means that if the user defines a hook that accepts `pageContext: PageContextServer | SomethingElse` then that's fine: everything will work. Maybe that's what TypeScript means with "function type A is assignable to function type B".

Reply:
[You seem to be right](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgApwOYQMIHtwQAeYAytAG7TIDeAUMmphAFzIDOYUoG9yC+MYBlYcuIHgxi5cIzt1oBfWrVCRYiFOix4CxbABtgEcDV5aW7OeN78Qg4ZbETkAIzhRZTxcrABPAA4oAPIgAEIQUlAQAErGACbQAGIAriAIYMD4yAC8yAAU-kw6kMSs5sVEpBTQAJQ5AHzI5LjAccq2HMj44ZEx8dCsIT24UbEgCVApaRlZuQVF+CVgZQu6VVCUUMgAPozai5UGRuA1rM2tDTRKQA).

So is it fine for you as it is now? I.e. `Config['onBeforeRender']` is defined to accept only `PageContextServer` but the user can still define a hook accepting `PageContextServer | PageContextClient`. I would cover that in the documentation.

Reply:
Sound good 👍